### PR TITLE
StyleGAN2 - 45836769

### PR DESCRIPTION
**Pull request by Hamo45746**

### CODE REVIEW

**Comment by Hamo45746** [comment]
Started work on the project on my own lab repo so there are some commits from that in the history. Ignore commits prior to [122ba3f].

**Comment by Uto0op** [comment]
Recognition problem: StyleGAN on ADNI ---- Hard Difficulty

1. Algorithm solves the problem: (3 Marks)
Poor generated image (black) -1;
No embedding plot (and a brief justification) -1

2. Algorithm demonstrated and functions: (2.5 Marks)
Algorithm demonstrated properly, yet failed to function as expected -0.5

4. Design: (1 Mark)
Modularized design

5. Commenting:
Sufficient docstrings and comments (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:
- Despite the unexpected poor functionality and results, it's good to record the attempts made to improve the performance. Some potential causes that I can think of:
  -- Not sure what LRs you tried, but the instability of both D/G training means small LRs 
  are desired;
  -- Try z_dim of 512 instead of 128, which could restraint the features learned;
  -- You mentioned progressive growing was used, but I don't see that in the code --- maybe was mistakenly (un)used and caused problem?
  -- The potential issue you mentioned could be possible;
  -- I would recommend starting with a basic StyleGAN, as long as it functions (i.e. 
  generates poor images with absurd losses), and then adding stuff like residual blocks 
  with skip connect, gradient clipping, adaptive discriminator augmentation to improve the 
  performance ---- at least you get a functioning model this way.

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: Nice documentation and introduction of the network. As shown in the loss plots, the model was not yet converged, and the generated image was just nothing. Nice analysis of debugging.  
2. **Implementation**: Well structured
3. **Design**: Good
4. **Comment**: Basic comments, could have more detailed comments like input and output of functions.
5. **Difficulty**: Hard

**Note**
One important feature of StyleGAN2 is Demodulation Instead of Normalization, group normalization would likely interfere with the benefits of weight demodulation. A better and deeper understanding of StyleGAN2 will help you.

**Comment by hanemma7moud** [comment]
Hamish Macintosh

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files (gitignore). Please update for merge, doesn't affect grade.
